sdk/storage/Azure.Storage.Blobs/src/RollingBufferStream.cs

```diff
@@ -0,0 +1,213 @@
+
+        /// <summary>
+        /// Index of the byte reserved for splitting the buffer "start" from "end".
+        /// </summary>
+        private int _reservedBytePos = 0;
+
+        /// <summary>
+        /// Whether the buffer has been filled enough to begin rolling.
+        /// </summary>
+        private bool _hasFilled = false;
+
+        /// <summary>
+        /// Size of the buffer in bytes.
+        /// </summary>
+        public int BufferSize => _rollingBuffer.Length;
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => true;
+
+        public override bool CanWrite => false;
+
+        /// <summary>
+        /// Expected length of the underlying stream.
+        /// </summary>
+        /// <remarks>
+        /// This class is to wrap unseekable streams, which, by nature, do not have a length to read.
+        /// </remarks>
+        private long? _expectedLength;
+        public override long Length => _expectedLength ?? _underlyingStreamBytesRead;
+
+        public override long Position
+        {
+            // easy to calculate position of stream if we haven't rolled over buffer
+            // otherwise, math based on buffer position and counter on underlying stream
+            get => _hasFilled
+                ? _underlyingStreamBytesRead - ((_reservedBytePos - _nextBufferBytePos + BufferSize) % BufferSize)
+                : _nextBufferBytePos;
+            // set read position within rolling buffer
+            set
+            {
+                if (value < 0)
+                {
+                    throw new ArgumentException("Cannot seek to a negative position");
+                }
+                if (value < _underlyingStreamBytesRead - BufferSize + 1)
+                {
+                    throw new ArgumentException("Tried to seek back further than buffer allowed.");
+                }
+                if (value > _underlyingStreamBytesRead)
+                {
+                    throw new ArgumentException("Tried to seek ahead of what has already been read from stream");
+                }
+
+                _nextBufferBytePos = (int)(value % BufferSize);
+            }
+        }
+
+        public RollingBufferStream(Stream stream, int bufferSize, long? expectedLength = default)
+        {
+            this._underlyingStream = stream;
+            this._rollingBuffer = new byte[bufferSize];
+            this._expectedLength = expectedLength;
+        }
+
+        public override void Flush()
+        {
+            /* https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.flush
+             * In a class derived from Stream that doesn't support writing, Flush is typically implemented
+             * as an empty method to ensure full compatibility with other Stream types since it's valid to
+             * flush a read-only stream.
+             */
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
```

Comment:
We need an async implementation too or you'll introduce blocking into async code paths _(the base ReadAsync will call into Read if it hasn't been overridden)_.  Ideally we'd share it with a `bool async` helper.